src/parser.js

```diff
@@ -1148,6 +1156,19 @@ function mapSwitchStatement(node, meta) {
 }
 
 function mapForStatement(node, meta) {
+  let blockStatement = mapBlockStatement(node.body, meta);
+
+  if (node.guard) {
+    const guardBase = node.guard.expression || node.guard;
+
+    blockStatement = b.blockStatement([
+      b.ifStatement(
+        mapOp(guardBase, meta),
```

Comment:
My example was more for the complex condition part, not the `foo for foo...` part. Agreed that other tests show this is not a problem 👍